### PR TITLE
Move json middleware (request and response) from faraday_middleware

### PR DESCRIPTION
**Pull request by iMacTia**
## Description

The JSON middleware, both request and response ones, are between the most used and require no external dependency since the `json` module is part of Ruby.
For this reason, we can safely move it inside Faraday.

Tests have also been moved and the Website documentation have been updated.

## Todos
List any remaining work that needs to be done, i.e:
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by olleolleolle** [comment]
Ruby 3 seems to have some unwanted behaviour about arguments, is that around kwargs?

**Comment by iMacTia** [comment]
It's the dependency loader, even adding `**` to the middleware initializer in the spec doesn't solve the problem completely. I was planning to get rid of it anyway so I'll do that in this PR

**Comment by iMacTia** [comment]
I'm gonna finish this up once #1301 gets merged 👍

File: docs/middleware/list.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ base64 representation.
 * [`Multipart`][multipart] converts a `Faraday::Request#body` hash of key/value pairs into a
 multipart form request.
 * [`UrlEncoded`][url_encoded] converts a `Faraday::Request#body` hash of key/value pairs into a url-encoded request body.
+* [`Json Request`][json-request] converts a `Faraday::Request#body` hash of key/value pairs into a json request body.
```
Comment (olleolleolle) [inline docs/middleware/list.md:30]: We say JSON in uppercase.

```suggestion
* [`Json Request`][json-request] converts a `Faraday::Request#body` hash of key/value pairs into a JSON request body.
```

File: docs/middleware/request/instrumentation.md [hunk 2]
```diff
@@ -5,8 +5,8 @@ permalink: /middleware/instrumentation
 hide: true
 prev_name: Retry Middleware
 prev_link: ./retry
-next_name: Logger Middleware
-next_link: ./logger
+next_name: Json Response Middleware
```
Comment (olleolleolle) [inline docs/middleware/request/instrumentation.md:8]: Perhaps uppercase, here?

```suggestion
next_name: JSON Response Middleware
```

File: docs/middleware/request/json.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+---
+layout: documentation
+title: "Json Request Middleware"
```
Comment (olleolleolle) [inline docs/middleware/request/json.md:3]:
```suggestion
title: "JSON Request Middleware"
```

File: docs/middleware/request/json.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+hide: true
+prev_name: UrlEncoded Middleware
+prev_link: ./url-encoded
+next_name: Retry Middleware
+next_link: ./retry
+top_name: Back to Middleware
+top_link: ./list
+---
+
+The `Json Request` middleware converts a `Faraday::Request#body` hash of key/value pairs into a json request body.
```
Comment (olleolleolle) [inline docs/middleware/request/json.md:14]:
```suggestion
The `Json Request` middleware converts a `Faraday::Request#body` hash of key/value pairs into a JSON request body.
```

File: docs/middleware/request/json.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+The middleware also automatically sets the `Content-Type` header to `application/json`,
+processes only requests with matching Content-type or those without a type and
```
Comment (olleolleolle) [inline docs/middleware/request/json.md:16]:
```suggestion
processes only requests with matching Content-Type or those without a type and
```

File: docs/middleware/request/url_encoded.md [hunk 6]
```diff
@@ -5,8 +5,8 @@ permalink: /middleware/url-encoded
 hide: true
 prev_name: Multipart Middleware
 prev_link: ./multipart
-next_name: Retry Middleware
-next_link: ./retry
+next_name: Json Request Middleware
```
Comment (olleolleolle) [inline docs/middleware/request/url_encoded.md:8]:
```suggestion
next_name: JSON Request Middleware
```

File: docs/middleware/response/json.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+---
+layout: documentation
+title: "Json Response Middleware"
```
Comment (olleolleolle) [inline docs/middleware/response/json.md:3]:
```suggestion
title: "JSON Response Middleware"
```

File: docs/middleware/response/json.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+hide: true
+prev_name: Instrumentation Middleware
+prev_link: ./instrumentation
+next_name: Logger Middleware
+next_link: ./logger
+top_name: Back to Middleware
+top_link: ./list
+---
+
+The `Json Response` middleware parses response body into a hash of key/value pairs.
```
Comment (olleolleolle) [inline docs/middleware/response/json.md:14]: Should this be referred  to like Json::Response here?

File: docs/middleware/response/json.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+The behaviour can be customized with the following options:
+* **parser_options:** options that will be sent to the JSON.parse method. Defaults to {}
+* **content_type:** Single value or Array of response content-types that should be processed. Can be either strings or Regex. Defaults to `/\bjson$/`
+* **preserve_raw:** If set to true, the original un-parsed response will be stored in the `response.env[:raw_body]` property. Defaults to `false`
```
Comment (olleolleolle) [inline docs/middleware/response/json.md:18]: These should have a period at the end of sentences.

File: docs/middleware/response/logger.md [hunk 10]
```diff
@@ -3,8 +3,8 @@ layout: documentation
 title: "Logger Middleware"
 permalink: /middleware/logger
 hide: true
-prev_name: Instrumentation Middleware
-prev_link: ./instrumentation
+prev_name: Json Response Middleware
```
Comment (olleolleolle) [inline docs/middleware/response/logger.md:6]:
```suggestion
prev_name: JSON Response Middleware
```

File: lib/faraday/request.rb [hunk 11]
```diff
@@ -140,11 +141,11 @@ def marshal_dump
...
-      self.body        = serialised[:body]
-      self.headers     = serialised[:headers]
-      self.path        = serialised[:path]
-      self.params      = serialised[:params]
-      self.options     = serialised[:options]
+      self.body = serialised[:body]
+      self.headers = serialised[:headers]
+      self.path = serialised[:path]
+      self.params = serialised[:params]
+      self.options = serialised[:options]
```
Comment (olleolleolle) [inline lib/faraday/request.rb:130]: I like this change!

File: lib/faraday/request/json.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      def match_content_type(env)
+        return unless process_request?(env)
+
+        env[:request_headers][CONTENT_TYPE] ||= MIME_TYPE
+        yield env[:body] unless env[:body].respond_to?(:to_str)
+      end
+
+      def process_request?(env)
+        type = request_type(env)
+        body?(env) && (type.empty? || MIME_TYPE_REGEX =~ type)
```
Comment (olleolleolle) [inline lib/faraday/request/json.rb:39]: Ruby version question: Can we use `#match?` here?

File: lib/faraday/response/json.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+      def parse_response?(env)
+        process_response_type?(env) &&
+          env[:body].respond_to?(:to_str)
+      end
+
+      def process_response_type?(env)
+        type = response_type(env)
+        @content_types.empty? || @content_types.any? do |pattern|
+          pattern.is_a?(Regexp) ? type =~ pattern : type == pattern
```
Comment (olleolleolle) [inline lib/faraday/response/json.rb:41]: Here, perhaps too, an opportunity for `#match?`

File: spec/faraday/response/json_spec.rb [hunk 14]
```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+RSpec.describe Faraday::Response::Json, type: :response do
+  let(:options) { {} }
+  let(:headers) { {} }
+  let(:middleware) do
+    puts options
```
Comment (olleolleolle) [inline spec/faraday/response/json_spec.rb:7]: This is debug output, perhaps it can be removed?

**Review by olleolleolle — COMMENTED** [review]
Wonderful!

File: docs/middleware/response/json.md (hunk 8 above)
Comment (iMacTia) [reply docs/middleware/response/json.md:14]: The class is technically `Faraday::Response::Json`, I could go with that or just `JSON` and move response outside as an adjective to "middleware"

Comment (iMacTia) [reply docs/middleware/response/json.md]: I went for the latter (and applied the same change to the request one), but happy to change if you prefer the other

**Review by olleolleolle — APPROVED** [review]
Looks good to me!

**Pull request merged**